// parse-server/lib/TestUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Connections = void 0;
exports.destroyAllDataPermanently = destroyAllDataPermanently;
exports.getConnectionsCount = getConnectionsCount;
exports.resolvingPromise = resolvingPromise;
exports.sleep = sleep;
var _cache = _interopRequireDefault(require("./cache"));
var _SchemaCache = _interopRequireDefault(require("./Adapters/Cache/SchemaCache"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Destroys all data in the database
 * @param {boolean} fast set to true if it's ok to just drop objects and not indexes.
 */
function destroyAllDataPermanently(fast) {
  if (!process.env.TESTING) {
    throw 'Only supported in test environment';
  }
  return Promise.all(Object.keys(_cache.default.cache).map(appId => {
    const app = _cache.default.get(appId);
    const deletePromises = [];
    if (app.cacheAdapter && app.cacheAdapter.clear) {
      deletePromises.push(app.cacheAdapter.clear());
    }
    if (app.databaseController) {
      deletePromises.push(app.databaseController.deleteEverything(fast));
    } else if (app.databaseAdapter) {
      _SchemaCache.default.clear();
      deletePromises.push(app.databaseAdapter.deleteAllClasses(fast));
    }
    return Promise.all(deletePromises);
  }));
}
function resolvingPromise() {
  let res;
  let rej;
  const promise = new Promise((resolve, reject) => {
    res = resolve;
    rej = reject;
  });
  promise.resolve = res;
  promise.reject = rej;
  return promise;
}
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
function getConnectionsCount(server) {
  return new Promise((resolve, reject) => {
    server.getConnections((err, count) => {
      /* istanbul ignore next */
      if (err) {
        reject(err);
      } else {
        resolve(count);
      }
    });
  });
}
;
class Connections {
  constructor() {
    this.sockets = new Set();
  }
  track(server) {
    server.on('connection', socket => {
      this.sockets.add(socket);
      socket.on('close', () => {
        this.sockets.delete(socket);
      });
    });
  }
  destroyAll() {
    for (const socket of this.sockets.values()) {
      socket.destroy();
    }
    this.sockets.clear();
  }
  count() {
    return this.sockets.size;
  }
}
exports.Connections = Connections;
